File: src/solutions/day4.clj
(ns solutions.day4)

(defn part1-solutions [] 
   (for [password (range 171309 643604)
         :let [lpw (map #(Character/digit % 10) (str password))]
         :when (every? (fn [[x y]] (>= y x)) (partition 2 1 lpw))
         :when (some (fn [[x y]] (= x y)) (partition 2 1 lpw))]
     lpw))

(println (count (part1-solutions)))

(defn part2-solutions []
  (filter #(some (fn [[_ y]] (= y 2)) (frequencies %)) (part1-solutions)))

(println (count (part2-solutions)))

File: src/solutions/day4.py
def only_sorted():
    pwl = [[int(j) for j in str(i)] for i in range(171309, 643604)]
    return list(filter(lambda x: x == sorted(x), pwl))

def part1():
    pwl = only_sorted()
    return list(filter(lambda x: any(freq > 1 for freq in [x.count(freq) for freq in x]), pwl))

print("part1 possible passwords: ", len(part1()))

def part2():
    pwl = only_sorted()
    return list(filter(lambda x: any(freq == 2 for freq in [x.count(freq) for freq in x]), pwl))

print("part2 possible passwords: ", len(part2()))
